Crash on nic configuration change

---
name: Bug report
about: Crash on nic configuration change
labels: bug, new

---

Giving access to our whole setup is a bit complex and I did not try to recreate a simple test case, as the problem is related to a change in our setup.

We changed our instance setup to add an internal network with separated nic interfaces and private ips to be able to communicate between the instances without going outside. 

Each time I have changes on the nic interface itself I get the crash copied below. If I destroy the vm (thanks to Fedora CoreOS, this is easy to recreate) and the nic interface, everything is fine; no crash.

### Terraform Version

```
 tofu version
OpenTofu v1.9.0
on linux_amd64
+ provider registry.opentofu.org/carlpett/sops v1.1.1
+ provider registry.opentofu.org/hashicorp/aws v5.86.1
+ provider registry.opentofu.org/hashicorp/local v2.5.2
+ provider registry.opentofu.org/outscale/outscale v1.0.1
+ provider registry.opentofu.org/poseidon/ct v0.13.0
```

### Terraform Configuration Files

```terraform
resource "outscale_nic" "internal_nic" {
  description        = "${var.name} internal network interface"
  subnet_id          = var.network_info.internal_subnet_id
  security_group_ids = [outscale_security_group.security_group.security_group_id]
  private_ips {
    is_primary = true
    private_ip = local.instance_vars.private_ips.internal
  }
  tags {
    key   = "Name"
    value = "${var.name} internal"
  }
}

resource "outscale_nic" "external_nic" {
  description        = "${var.name} external network interface"
  subnet_id          = var.network_info.external_subnet_id
  security_group_ids = [outscale_security_group.security_group.security_group_id]
  private_ips {
    is_primary = true
    private_ip = local.instance_vars.private_ips.external
  }
  tags {
    key   = "Name"
    value = "${var.name} external"
  }
}

resource "outscale_vm" "vm" {
  image_id = var.image_id
  vm_type  = local.vm_type
  primary_nic {
    nic_id        = outscale_nic.external_nic.nic_id
    device_number = 0
  }

  tags {
    key   = "name"
    value = var.name
  }
  user_data = base64encode(terraform_data.vm_user_data.output)
  lifecycle {
    replace_triggered_by = [terraform_data.vm_user_data]
  }
}

resource "outscale_nic_link" "nic_link_internal" {
  nic_id        = outscale_nic.internal_nic.nic_id
  device_number = 1
  vm_id         = outscale_vm.vm.vm_id
}
```

### Crash Output
```
│ Error: Request cancelled
│
│   with module.fcos_instances["monitoring"].outscale_nic.external_nic,
│   on fcos/outscale.tf line 31, in resource "outscale_nic" "external_nic":
│   31: resource "outscale_nic" "external_nic" {
│
│ The plugin.(*GRPCProvider).ApplyResourceChange request was cancelled.
╵

Stack trace from the terraform-provider-outscale_v1.0.1 plugin:

panic: interface conversion: interface {} is []interface {}, not *schema.Set

goroutine 85 [running]:
github.com/outscale/terraform-provider-outscale/outscale.ResourceOutscaleNicUpdate(0xc0008fe600, {0x10fa780, 0xc00012abb8})
        github.com/outscale/terraform-provider-outscale/outscale/resource_outscale_nic.go:502 +0x2015
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0x15e8018?, {0x15e8018?, 0xc0008d3c50?}, 0xd?, {0x10fa780?, 0xc00012abb8?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:800 +0x15f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00045f5e0, {0x15e8018, 0xc0008d3c50}, 0xc00086bad0, 0xc0008fe180, {0x10fa780, 0xc00012abb8})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:919 +0x81a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000010018, {0x15e8018?, 0xc0008d3b90?}, 0xc0008addb0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/grpc_provider.go:1060 +0xd5c
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ApplyResourceChange(0xc000810000, {0x15e8018?, 0xc0008d38c0?}, 0xc0008addb0)
        github.com/hashicorp/terraform-plugin-mux@v0.12.0/tf5muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc00062e140, {0x15e8018?, 0xc0008d2f00?}, 0xc000537500)
        github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/tf5server/server.go:859 +0x548
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x12d6700, 0xc00062e140}, {0x15e8018, 0xc0008d2f00}, 0xc00084b800, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:467 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001b4c00, {0x15e8018, 0xc0008d2e70}, {0x15ee480, 0xc00080a340}, 0xc0008cf8c0, 0xc000670ff0, 0x1f57dd8, 0x0)
        google.golang.org/grpc@v1.62.0/server.go:1383 +0xdd1
google.golang.org/grpc.(*Server).handleStream(0xc0001b4c00, {0x15ee480, 0xc00080a340}, 0xc0008cf8c0)
        google.golang.org/grpc@v1.62.0/server.go:1794 +0xe87
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.62.0/server.go:1027 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 52
        google.golang.org/grpc@v1.62.0/server.go:1038 +0x125

Error: The terraform-provider-outscale_v1.0.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Pending changes 

```
  # module.fcos_instances["monitoring"].outscale_nic.external_nic will be updated in-place
  ~ resource "outscale_nic" "external_nic" {
        id                     = "eni-4a192814"
      ~ subnet_id              = "subnet-4b48d3ab" -> "subnet-4a265a67"
        # (13 unchanged attributes hidden)

      ~ private_ips {
          ~ private_ip       = "192.168.14.211" -> "192.168.16.12"
            # (3 unchanged attributes hidden)
        }

      + tags {
          + key   = "Name"
          + value = "monitoring external"
        }
```

--- a/outscale/resource_outscale_nic.go
+++ b/outscale/resource_outscale_nic.go
@@ -4,16 +4,14 @@ import (
 	"context"
 	"fmt"
 	"log"
-	"math"
 	"net/http"
 	"reflect"
 	"strconv"
-	"strings"
 	"time"
 
 	oscgo "github.com/outscale/osc-sdk-go/v2"
 
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/outscale/terraform-provider-outscale/utils"
 )
@@ -30,6 +28,8 @@ func ResourceOutscaleNic() *schema.Resource {
 		},
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(10 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(5 * time.Minute),
 			Delete: schema.DefaultTimeout(10 * time.Minute),
 		},
 		Schema: getOAPINicSchema(),
@@ -158,14 +158,13 @@ func getOAPINicSchema() map[string]*schema.Schema {
 		},
 
 		"private_ips": {
-			Type:     schema.TypeList,
+			Type:     schema.TypeSet,
 			Computed: true,
 			Optional: true,
-			ForceNew: true,
 			Elem: &schema.Resource{
 				Schema: map[string]*schema.Schema{
 					"link_public_ip": {
-						Type:     schema.TypeList,
+						Type:     schema.TypeSet,
 						Computed: true,
 						Elem: &schema.Resource{
 							Schema: map[string]*schema.Schema{
@@ -249,13 +248,11 @@ func ResourceOutscaleNicCreate(d *schema.ResourceData, meta interface{}) error {
 		request.SetSecurityGroupIds(sgIDs)
 	}
 	if v, ok := d.GetOk("private_ips"); ok {
-		request.SetPrivateIps(expandPrivateIPLight(v.([]interface{})))
+		request.SetPrivateIps(expandPrivateIPLight(v.(*schema.Set).List()))
 	}
 
-	log.Printf("[DEBUG] Creating network interface")
-
 	var resp oscgo.CreateNicResponse
-	err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+	err := retry.RetryContext(context.Background(), d.Timeout(schema.TimeoutCreate), func() *retry.RetryError {
 		var err error
 		rp, httpResp, err := conn.NicApi.CreateNic(context.Background()).CreateNicRequest(request).Execute()
 		if err != nil {
@@ -298,7 +295,7 @@ func ResourceOutscaleNicRead(d *schema.ResourceData, meta interface{}) error {
 
 	var resp oscgo.ReadNicsResponse
 	var err error
-	err = resource.Retry(5*time.Minute, func() *resource.RetryError {
+	err = retry.RetryContext(context.Background(), d.Timeout(schema.TimeoutRead), func() *retry.RetryError {
 		rp, httpResp, err := conn.NicApi.ReadNics(context.Background()).ReadNicsRequest(dnir).Execute()
 		if err != nil {
 			return utils.CheckThrottling(httpResp, err)
@@ -311,13 +308,6 @@ func ResourceOutscaleNicRead(d *schema.ResourceData, meta interface{}) error {
 		return fmt.Errorf("Error describing Network Interfaces : %s", err)
 	}
 
-	if err != nil {
-		if strings.Contains(err.Error(), "Unable to find Nic") {
-			d.SetId("")
-			return nil
-		}
-		return fmt.Errorf("Error retrieving ENI: %s", err)
-	}
 	if utils.IsResponseEmpty(len(resp.GetNics()), "Nic", d.Id()) {
 		d.SetId("")
 		return nil
@@ -387,10 +377,7 @@ func ResourceOutscaleNicRead(d *schema.ResourceData, meta interface{}) error {
 func ResourceOutscaleNicDelete(d *schema.ResourceData, meta interface{}) error {
 
 	conn := meta.(*OutscaleClient).OSCAPI
-
-	log.Printf("[INFO] Deleting ENI: %s", d.Id())
-
-	err := ResourceOutscaleNicDetach(meta, d.Id())
+	err := ResourceOutscaleNicDetach(meta, d.Id(), d.Timeout(schema.TimeoutDelete))
 	if err != nil {
 		return err
 	}
@@ -399,7 +386,7 @@ func ResourceOutscaleNicDelete(d *schema.ResourceData, meta interface{}) error {
 		NicId: d.Id(),
 	}
 
-	err = resource.Retry(5*time.Minute, func() *resource.RetryError {
+	err = retry.RetryContext(context.Background(), d.Timeout(schema.TimeoutDelete), func() *retry.RetryError {
 		_, httpResp, err := conn.NicApi.DeleteNic(context.Background()).DeleteNicRequest(deleteEniOpts).Execute()
 		if err != nil {
 			return utils.CheckThrottling(httpResp, err)
@@ -408,24 +395,24 @@ func ResourceOutscaleNicDelete(d *schema.ResourceData, meta interface{}) error {
 	})
 
 	if err != nil {
-
 		return fmt.Errorf("Error Deleting ENI: %s", err)
 	}
 
 	return nil
 }
 
-func ResourceOutscaleNicDetach(meta interface{}, nicID string) error {
+func ResourceOutscaleNicDetach(meta interface{}, nicID string, timeout time.Duration) error {
 	// if there was an old nic_link, remove it
 	conn := meta.(*OutscaleClient).OSCAPI
 
-	stateConf := &resource.StateChangeConf{
+	stateConf := &retry.StateChangeConf{
 		Pending: []string{"attaching", "detaching"},
 		Target:  []string{"attached", "detached", "failed"},
 		Refresh: nicLinkRefreshFunc(conn, nicID),
-		Timeout: 10 * time.Minute,
+		Timeout: timeout,
+		Delay:   1 * time.Second,
 	}
-	resp, err := stateConf.WaitForState()
+	resp, err := stateConf.WaitForStateContext(context.Background())
 	if err != nil {
 		return fmt.Errorf(
 			"Error waiting for ENI (%s) to become dettached: %s", nicID, err)
@@ -441,7 +428,7 @@ func ResourceOutscaleNicDetach(meta interface{}, nicID string) error {
 			LinkNicId: linkNic.GetLinkNicId(),
 		}
 		var statusCode int
-		err = resource.Retry(5*time.Minute, func() *resource.RetryError {
+		err := retry.RetryContext(context.Background(), timeout, func() *retry.RetryError {
 			_, httpResp, err := conn.NicApi.UnlinkNic(context.Background()).UnlinkNicRequest(req).Execute()
 			if err != nil {
 				return utils.CheckThrottling(httpResp, err)
@@ -464,48 +451,18 @@ func ResourceOutscaleNicUpdate(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*OutscaleClient).OSCAPI
 	var err error
 
-	if d.HasChange("link_nic") {
-		_, na := d.GetChange("link_nic")
-
-		err := ResourceOutscaleNicDetach(meta, d.Id())
-		if err != nil {
-			return err
-		}
-
-		// if there is a new nic_link, attach it
-		if na != nil && len(na.([]interface{})) > 0 {
-			na := na.([]interface{})[0].(map[string]interface{})
-			di := na["device_number"].(int)
-			ar := oscgo.LinkNicRequest{
-				DeviceNumber: int32(di),
-				VmId:         na["instance"].(string),
-				NicId:        d.Id(),
-			}
-
-			err = resource.Retry(5*time.Minute, func() *resource.RetryError {
-				_, httpResp, err := conn.NicApi.LinkNic(context.Background()).LinkNicRequest(ar).Execute()
-				if err != nil {
-					return utils.CheckThrottling(httpResp, err)
-				}
-				return nil
-			})
-			if err != nil {
-				return fmt.Errorf("Error Attaching Network Interface: %s", err)
-			}
-		}
-	}
-
 	if d.HasChange("private_ips") {
-		o, n := d.GetChange("private_ips")
+		oldIps, newIps := d.GetChange("private_ips")
+		removed, created := utils.GetTypeSetDifferencesForUpdating(oldIps.(*schema.Set), newIps.(*schema.Set))
 
 		// Unassign old IP addresses
-		if len(o.(*schema.Set).List()) != 0 {
+		if listOldIps := removed.List(); len(listOldIps) != 0 {
 			input := oscgo.UnlinkPrivateIpsRequest{
 				NicId:      d.Id(),
-				PrivateIps: flattenPrivateIPLightToStringSlice(o.(*schema.Set).List()),
+				PrivateIps: flattenPrivateIPLightToStringSlice(listOldIps),
 			}
 
-			err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+			err := retry.RetryContext(context.Background(), d.Timeout(schema.TimeoutUpdate), func() *retry.RetryError {
 				_, httpResp, err := conn.NicApi.UnlinkPrivateIps(context.Background()).UnlinkPrivateIpsRequest(input).Execute()
 				if err != nil {
 					return utils.CheckThrottling(httpResp, err)
@@ -518,14 +475,14 @@ func ResourceOutscaleNicUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 
 		// Assign new IP addresses
-		if len(n.(*schema.Set).List()) != 0 {
-			stringSlice := flattenPrivateIPLightToStringSlice(n.(*schema.Set).List())
+		if listNewIps := created.List(); len(listNewIps) != 0 {
+			stringSlice := flattenPrivateIPLightToStringSlice(listNewIps)
 			input := oscgo.LinkPrivateIpsRequest{
 				NicId:      d.Id(),
 				PrivateIps: &stringSlice,
 			}
 
-			err = resource.Retry(5*time.Minute, func() *resource.RetryError {
+			err = retry.RetryContext(context.Background(), d.Timeout(schema.TimeoutUpdate), func() *retry.RetryError {
 				_, httpResp, err := conn.NicApi.LinkPrivateIps(context.Background()).LinkPrivateIpsRequest(input).Execute()
 				if err != nil {
 					return utils.CheckThrottling(httpResp, err)
@@ -538,68 +495,13 @@ func ResourceOutscaleNicUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 	}
 
-	if d.HasChange("private_ips_count") {
-		o, n := d.GetChange("private_ips_count")
-		pips := d.Get("pips").(*schema.Set).List()
-		prips := pips[:0]
-		pip := d.Get("private_ip")
-
-		for _, ip := range pips {
-			if ip != pip {
-				prips = append(prips, ip)
-			}
-		}
-
-		if o != nil && o != 0 && n != nil && n != len(prips) {
-			diff := n.(int) - o.(int)
-
-			// Surplus of IPs, add the diff
-			if diff > 0 {
-				input := oscgo.LinkPrivateIpsRequest{
-					NicId: d.Id(),
-				}
-				input.SetSecondaryPrivateIpCount(int32(diff))
-
-				err := resource.Retry(5*time.Minute, func() *resource.RetryError {
-					var err error
-					_, httpResp, err := conn.NicApi.LinkPrivateIps(context.Background()).LinkPrivateIpsRequest(input).Execute()
-					if err != nil {
-						return utils.CheckThrottling(httpResp, err)
-					}
-					return nil
-				})
-				if err != nil {
-					return fmt.Errorf("Failure to assign Private IPs: %s", err)
-				}
-			}
-
-			if diff < 0 {
-				input := oscgo.UnlinkPrivateIpsRequest{
-					NicId:      d.Id(),
-					PrivateIps: utils.InterfaceSliceToStringSlice(prips[0:int(math.Abs(float64(diff)))]),
-				}
-
-				err := resource.Retry(5*time.Minute, func() *resource.RetryError {
-					_, httpResp, err := conn.NicApi.UnlinkPrivateIps(context.Background()).UnlinkPrivateIpsRequest(input).Execute()
-					if err != nil {
-						return utils.CheckThrottling(httpResp, err)
-					}
-					return nil
-				})
-				if err != nil {
-					return fmt.Errorf("Failure to unassign Private IPs: %s", err)
-				}
-			}
-		}
-	}
-
 	if d.HasChange("security_group_ids") {
 		request := oscgo.UpdateNicRequest{
 			NicId: d.Id(),
 		}
 		request.SetSecurityGroupIds(utils.SetToStringSlice(d.Get("security_group_ids").(*schema.Set)))
 
-		err = resource.Retry(5*time.Minute, func() *resource.RetryError {
+		err = retry.RetryContext(context.Background(), d.Timeout(schema.TimeoutUpdate), func() *retry.RetryError {
 			_, httpResp, err := conn.NicApi.UpdateNic(context.Background()).UpdateNicRequest(request).Execute()
 			if err != nil {
 				return utils.CheckThrottling(httpResp, err)
@@ -618,7 +520,7 @@ func ResourceOutscaleNicUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 		request.SetDescription(d.Get("description").(string))
 		var err error
-		err = resource.Retry(5*time.Minute, func() *resource.RetryError {
+		err = retry.RetryContext(context.Background(), d.Timeout(schema.TimeoutUpdate), func() *retry.RetryError {
 			_, httpResp, err := conn.NicApi.UpdateNic(context.Background()).UpdateNicRequest(request).Execute()
 			if err != nil {
 				return utils.CheckThrottling(httpResp, err)

--- a/outscale/resource_outscale_nic_test.go
+++ b/outscale/resource_outscale_nic_test.go
@@ -2,14 +2,11 @@ package outscale
 
 import (
 	"fmt"
-	"reflect"
 	"testing"
 
-	oscgo "github.com/outscale/osc-sdk-go/v2"
 	"github.com/outscale/terraform-provider-outscale/utils"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 )
 
 func TestAccNet_WithNic_basic(t *testing.T) {
@@ -37,21 +34,6 @@ func TestAccNet_WithNic_basic(t *testing.T) {
 			},
 		},
 	})
-}
-
-func testAccCheckOutscaleENIAttributes(conf *oscgo.Nic, subregion string) resource.TestCheckFunc {
-	return func(s *terraform.State) error {
-
-		if !reflect.DeepEqual(conf.GetLinkNic(), oscgo.LinkNic{}) {
-			return fmt.Errorf("expected attachment to be nil")
-		}
-
-		if conf.GetSubregionName() != fmt.Sprintf("%sa", subregion) {
-			return fmt.Errorf("expected subregion_name to be %sa, but was %s", subregion, conf.GetSubregionName())
-		}
-
-		return nil
-	}
 }
 
 func testAccOutscaleENIConfig(subregion string) string {
@@ -105,7 +87,7 @@ func testAccOutscaleENIConfigUpdate(subregion string) string {
 
 		resource "outscale_subnet" "outscale_subnet" {
 			subregion_name = "%sa"
-			ip_range       = "10.0.0.0/24"
+			ip_range       = "10.0.0.0/16"
 			net_id         = outscale_net.outscale_net.net_id
 		}
 

--- a/outscale/resource_outscale_vm_test.go
+++ b/outscale/resource_outscale_vm_test.go
@@ -165,12 +165,12 @@ func TestAccVM_withTags(t *testing.T) {
 			{
 				Config: testAccVmsConfigUpdateOAPIVMTags(omi, utils.TestAccVmType, utils.GetRegion(), tagsValue, keypair),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckOutscaleVMExists("outscale_vm.basic", &server),
+					testAccCheckOutscaleVMExists("outscale_vm.basic_tags", &server),
 					testAccCheckOutscaleVMAttributes(t, &server, omi),
 					resource.TestCheckResourceAttr(
-						"outscale_vm.basic", "image_id", omi),
+						"outscale_vm.basic_tags", "image_id", omi),
 					resource.TestCheckResourceAttr(
-						"outscale_vm.basic", "vm_type", utils.TestAccVmType),
+						"outscale_vm.basic_tags", "vm_type", utils.TestAccVmType),
 				),
 			},
 		},
@@ -180,7 +180,7 @@ func TestAccVM_withTags(t *testing.T) {
 func TestAccNet_VM_withNics(t *testing.T) {
 	omi := os.Getenv("OUTSCALE_IMAGEID")
 	keypair := "terraform-basic"
-	resourceName := "outscale_vm.basic"
+	resourceName := "outscale_vm.basic_with_nic"
 
 	resource.Test(t, resource.TestCase{
 		PreCheck:                 func() { testAccPreCheck(t) },
@@ -727,7 +727,8 @@ func testAccCheckOutscaleVMConfigBasicWithNicAttached(omi, vmType, region, keypa
 }
 
 func testAccCheckOutscaleVMConfigBasicWithNics(omi, vmType, keypair, region string) string {
-	return fmt.Sprintf(`resource "outscale_net" "outscale_net" {
+	return fmt.Sprintf(`
+	  resource "outscale_net" "outscale_net" {
 		ip_range = "10.0.0.0/16"
 	  }
 
@@ -743,7 +744,7 @@ func testAccCheckOutscaleVMConfigBasicWithNics(omi, vmType, keypair, region stri
 		net_id              = outscale_net.outscale_net.net_id
 	  }
 
-	  resource "outscale_vm" "basic" {
+	  resource "outscale_vm" "basic_with_nic" {
 		image_id     = "%[1]s"
 		vm_type      = "%[2]s"
 		keypair_name = "%[3]s"
@@ -824,7 +825,7 @@ func testAccVmsConfigUpdateOAPIVMTags(omi, vmType, region, value, keypair string
 			description                  = "testAcc Terraform security group"
 			security_group_name          = "sgTagsVm"
 		}
-		resource "outscale_vm" "basic" {
+		resource "outscale_vm" "basic_tags" {
 			image_id                 = "%[1]s"
 			vm_type                  = "%[2]s"
 			keypair_name             = "%[5]s"

--- a/tests/qa_provider_oapi/data/load_balancer_vm_health/TF-92_load_balancer_vm_health_datasource_attributes_ok/step1.load_balancer_vm_health_datasource_attributes_ok.ref
+++ b/tests/qa_provider_oapi/data/load_balancer_vm_health/TF-92_load_balancer_vm_health_datasource_attributes_ok/step1.load_balancer_vm_health_datasource_attributes_ok.ref
@@ -216,7 +216,7 @@
                         "backend_vm_ids": [
                             "##id-0##"
                         ],
-                        "id": "##id-5##",
+                        "id": "##id-3##",
                         "load_balancer_name": "########",
                         "request_id": "########"
                     },
@@ -240,15 +240,15 @@
                 {
                     "schema_version": 0,
                     "attributes": {
-                        "account_id": "##id-6##",
+                        "account_id": "##id-5##",
                         "description": "test lbu vm health",
-                        "id": "##id-7##",
+                        "id": "##id-6##",
                         "inbound_rules": [],
                         "net_id": "",
                         "outbound_rules": [],
                         "remove_default_outbound_rule": false,
                         "request_id": "########",
-                        "security_group_id": "##id-7##",
+                        "security_group_id": "##id-6##",
                         "security_group_name": "sgTF-92",
                         "tag": null,
                         "tags": []
@@ -269,14 +269,14 @@
                     "attributes": {
                         "flow": "Inbound",
                         "from_port_range": 8080,
-                        "id": "##id-7##",
+                        "id": "##id-6##",
                         "ip_protocol": "tcp",
                         "ip_range": "0.0.0.0/0",
                         "net_id": "",
                         "request_id": "########",
                         "rules": [],
                         "security_group_account_id_to_link": null,
-                        "security_group_id": "##id-7##",
+                        "security_group_id": "##id-6##",
                         "security_group_name": "sgTF-92",
                         "security_group_name_to_link": null,
                         "to_port_range": 8080
@@ -309,7 +309,7 @@
                                         "link_date": "########",
                                         "state": "attached",
                                         "tags": [],
-                                        "volume_id": "##id-8##"
+                                        "volume_id": "##id-7##"
                                     }
                                 ],
                                 "device_name": "/dev/sda1"
@@ -322,7 +322,7 @@
                         "get_admin_password": null,
                         "hypervisor": "xen",
                         "id": "##id-0##",
-                        "image_id": "##id-9##",
+                        "image_id": "##id-8##",
                         "is_source_dest_checked": false,
                         "keypair_name": "KP-TF92",
                         "launch_number": 0,
@@ -343,16 +343,16 @@
                         "public_dns_name": "########",
                         "public_ip": "########",
                         "request_id": "########",
-                        "reservation_id": "##id-10##",
+                        "reservation_id": "##id-9##",
                         "root_device_name": "/dev/sda1",
                         "root_device_type": "ebs",
                         "security_group_ids": [
-                            "##id-7##"
+                            "##id-6##"
                         ],
                         "security_group_names": null,
                         "security_groups": [
                             {
-                                "security_group_id": "##id-7##",
+                                "security_group_id": "##id-6##",
                                 "security_group_name": "sgTF-92"
                             }
                         ],

--- a/tests/qa_provider_oapi/data/nets/TF-109_nic_datasource_attributes_ok/step1.nic_datasource_attributes_ok.ref
+++ b/tests/qa_provider_oapi/data/nets/TF-109_nic_datasource_attributes_ok/step1.nic_datasource_attributes_ok.ref
@@ -406,13 +406,13 @@
                         "private_ip": "########",
                         "private_ips": [
                             {
-                                "is_primary": true,
+                                "is_primary": false,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
                             },
                             {
-                                "is_primary": false,
+                                "is_primary": true,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
@@ -445,7 +445,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.security_group01",
@@ -476,13 +476,13 @@
                         "private_ip": "########",
                         "private_ips": [
                             {
-                                "is_primary": true,
+                                "is_primary": false,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
                             },
                             {
-                                "is_primary": false,
+                                "is_primary": true,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
@@ -511,7 +511,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.security_group01",
@@ -575,7 +575,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.security_group01",

--- a/tests/qa_provider_oapi/data/nets/TF-110_nic_resource_attributes_ok/step1.nic_resource_attributes_ok.ref
+++ b/tests/qa_provider_oapi/data/nets/TF-110_nic_resource_attributes_ok/step1.nic_resource_attributes_ok.ref
@@ -74,7 +74,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_subnet.outscale_subnet"
@@ -104,13 +104,13 @@
                         "private_ip": "########",
                         "private_ips": [
                             {
-                                "is_primary": true,
+                                "is_primary": false,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
                             },
                             {
-                                "is_primary": false,
+                                "is_primary": true,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
@@ -134,7 +134,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.outscale_sg",

--- a/tests/qa_provider_oapi/data/nets/TF-110_nic_resource_attributes_ok/step2.nic_resource_attributes_update.ref
+++ b/tests/qa_provider_oapi/data/nets/TF-110_nic_resource_attributes_ok/step2.nic_resource_attributes_update.ref
@@ -74,7 +74,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.outscale_sg",
@@ -105,13 +105,13 @@
                         "private_ip": "########",
                         "private_ips": [
                             {
-                                "is_primary": true,
+                                "is_primary": false,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
                             },
                             {
-                                "is_primary": false,
+                                "is_primary": true,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
@@ -135,7 +135,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.outscale_sg",

--- a/tests/qa_provider_oapi/data/nets/TF-111_nic_link_resource_attributes_ok/step1.nic_link_resource_attributes_ok.ref
+++ b/tests/qa_provider_oapi/data/nets/TF-111_nic_link_resource_attributes_ok/step1.nic_link_resource_attributes_ok.ref
@@ -96,7 +96,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.outscale_security_group",

--- a/tests/qa_provider_oapi/data/nets/TF-112_nic_private_ip_resource_attributes_ok/step1.nic_private_ip_resource_attributes_ok.ref
+++ b/tests/qa_provider_oapi/data/nets/TF-112_nic_private_ip_resource_attributes_ok/step1.nic_private_ip_resource_attributes_ok.ref
@@ -74,7 +74,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_subnet.outscale_subnet"

--- a/tests/qa_provider_oapi/data/nets/TF-113_nics_datasource_attributes_ok/step1.nics_datasource_attributes_ok.ref
+++ b/tests/qa_provider_oapi/data/nets/TF-113_nics_datasource_attributes_ok/step1.nics_datasource_attributes_ok.ref
@@ -254,6 +254,82 @@
                         "nics": [
                             {
                                 "account_id": "##id-1##",
+                                "description": "Primary network interface",
+                                "is_source_dest_checked": true,
+                                "link_nic": [
+                                    {
+                                        "delete_on_vm_deletion": true,
+                                        "device_number": 0,
+                                        "link_nic_id": "##id-9##",
+                                        "state": "attached",
+                                        "vm_account_id": "##id-1##",
+                                        "vm_id": "##id-3##"
+                                    }
+                                ],
+                                "link_public_ip": [],
+                                "mac_address": "########",
+                                "net_id": "##id-4##",
+                                "nic_id": "##id-10##",
+                                "private_dns_name": "########",
+                                "private_ips": [
+                                    {
+                                        "is_primary": true,
+                                        "link_public_ip": [],
+                                        "private_dns_name": "########",
+                                        "private_ip": "########"
+                                    }
+                                ],
+                                "security_groups": [
+                                    {
+                                        "security_group_id": "##id-6##",
+                                        "security_group_name": "Terraform-SG-TF113"
+                                    }
+                                ],
+                                "state": "in-use",
+                                "subnet_id": "##id-7##",
+                                "subregion_name": "us-east-2a",
+                                "tags": []
+                            },
+                            {
+                                "account_id": "##id-1##",
+                                "description": "",
+                                "is_source_dest_checked": true,
+                                "link_nic": [],
+                                "link_public_ip": [],
+                                "mac_address": "########",
+                                "net_id": "##id-4##",
+                                "nic_id": "##id-12##",
+                                "private_dns_name": "########",
+                                "private_ips": [
+                                    {
+                                        "is_primary": true,
+                                        "link_public_ip": [],
+                                        "private_dns_name": "########",
+                                        "private_ip": "########"
+                                    }
+                                ],
+                                "security_groups": [
+                                    {
+                                        "security_group_id": "##id-6##",
+                                        "security_group_name": "Terraform-SG-TF113"
+                                    }
+                                ],
+                                "state": "available",
+                                "subnet_id": "##id-13##",
+                                "subregion_name": "us-east-2b",
+                                "tags": [
+                                    {
+                                        "key": "Key:",
+                                        "value": ":value-tags"
+                                    },
+                                    {
+                                        "key": "Key-2",
+                                        "value": "value-tags-2"
+                                    }
+                                ]
+                            },
+                            {
+                                "account_id": "##id-1##",
                                 "description": "TF-113",
                                 "is_source_dest_checked": true,
                                 "link_nic": [
@@ -309,44 +385,6 @@
                                 "account_id": "##id-1##",
                                 "description": "",
                                 "is_source_dest_checked": true,
-                                "link_nic": [],
-                                "link_public_ip": [],
-                                "mac_address": "########",
-                                "net_id": "##id-4##",
-                                "nic_id": "##id-12##",
-                                "private_dns_name": "########",
-                                "private_ips": [
-                                    {
-                                        "is_primary": true,
-                                        "link_public_ip": [],
-                                        "private_dns_name": "########",
-                                        "private_ip": "########"
-                                    }
-                                ],
-                                "security_groups": [
-                                    {
-                                        "security_group_id": "##id-6##",
-                                        "security_group_name": "Terraform-SG-TF113"
-                                    }
-                                ],
-                                "state": "available",
-                                "subnet_id": "##id-13##",
-                                "subregion_name": "us-east-2b",
-                                "tags": [
-                                    {
-                                        "key": "Key-2",
-                                        "value": "value-tags-2"
-                                    },
-                                    {
-                                        "key": "Key:",
-                                        "value": ":value-tags"
-                                    }
-                                ]
-                            },
-                            {
-                                "account_id": "##id-1##",
-                                "description": "",
-                                "is_source_dest_checked": true,
                                 "link_nic": [
                                     {
                                         "delete_on_vm_deletion": false,
@@ -391,44 +429,6 @@
                                         "value": "Nic-2"
                                     }
                                 ]
-                            },
-                            {
-                                "account_id": "##id-1##",
-                                "description": "Primary network interface",
-                                "is_source_dest_checked": true,
-                                "link_nic": [
-                                    {
-                                        "delete_on_vm_deletion": true,
-                                        "device_number": 0,
-                                        "link_nic_id": "##id-9##",
-                                        "state": "attached",
-                                        "vm_account_id": "##id-1##",
-                                        "vm_id": "##id-3##"
-                                    }
-                                ],
-                                "link_public_ip": [],
-                                "mac_address": "########",
-                                "net_id": "##id-4##",
-                                "nic_id": "##id-10##",
-                                "private_dns_name": "########",
-                                "private_ips": [
-                                    {
-                                        "is_primary": true,
-                                        "link_public_ip": [],
-                                        "private_dns_name": "########",
-                                        "private_ip": "########"
-                                    }
-                                ],
-                                "security_groups": [
-                                    {
-                                        "security_group_id": "##id-6##",
-                                        "security_group_name": "Terraform-SG-TF113"
-                                    }
-                                ],
-                                "state": "in-use",
-                                "subnet_id": "##id-7##",
-                                "subregion_name": "us-east-2a",
-                                "tags": []
                             }
                         ],
                         "request_id": "########"
@@ -451,6 +451,82 @@
                         "nics": [
                             {
                                 "account_id": "##id-1##",
+                                "description": "Primary network interface",
+                                "is_source_dest_checked": true,
+                                "link_nic": [
+                                    {
+                                        "delete_on_vm_deletion": true,
+                                        "device_number": 0,
+                                        "link_nic_id": "##id-9##",
+                                        "state": "attached",
+                                        "vm_account_id": "##id-1##",
+                                        "vm_id": "##id-3##"
+                                    }
+                                ],
+                                "link_public_ip": [],
+                                "mac_address": "########",
+                                "net_id": "##id-4##",
+                                "nic_id": "##id-10##",
+                                "private_dns_name": "########",
+                                "private_ips": [
+                                    {
+                                        "is_primary": true,
+                                        "link_public_ip": [],
+                                        "private_dns_name": "########",
+                                        "private_ip": "########"
+                                    }
+                                ],
+                                "security_groups": [
+                                    {
+                                        "security_group_id": "##id-6##",
+                                        "security_group_name": "Terraform-SG-TF113"
+                                    }
+                                ],
+                                "state": "in-use",
+                                "subnet_id": "##id-7##",
+                                "subregion_name": "us-east-2a",
+                                "tags": []
+                            },
+                            {
+                                "account_id": "##id-1##",
+                                "description": "",
+                                "is_source_dest_checked": true,
+                                "link_nic": [],
+                                "link_public_ip": [],
+                                "mac_address": "########",
+                                "net_id": "##id-4##",
+                                "nic_id": "##id-12##",
+                                "private_dns_name": "########",
+                                "private_ips": [
+                                    {
+                                        "is_primary": true,
+                                        "link_public_ip": [],
+                                        "private_dns_name": "########",
+                                        "private_ip": "########"
+                                    }
+                                ],
+                                "security_groups": [
+                                    {
+                                        "security_group_id": "##id-6##",
+                                        "security_group_name": "Terraform-SG-TF113"
+                                    }
+                                ],
+                                "state": "available",
+                                "subnet_id": "##id-13##",
+                                "subregion_name": "us-east-2b",
+                                "tags": [
+                                    {
+                                        "key": "Key:",
+                                        "value": ":value-tags"
+                                    },
+                                    {
+                                        "key": "Key-2",
+                                        "value": "value-tags-2"
+                                    }
+                                ]
+                            },
+                            {
+                                "account_id": "##id-1##",
                                 "description": "TF-113",
                                 "is_source_dest_checked": true,
                                 "link_nic": [
@@ -493,50 +569,12 @@
                                 "subregion_name": "us-east-2a",
                                 "tags": [
                                     {
-                                        "key": "Key:",
-                                        "value": ":value-tags"
-                                    },
-                                    {
                                         "key": "Key-2",
                                         "value": "value-tags-2"
-                                    }
-                                ]
-                            },
-                            {
-                                "account_id": "##id-1##",
-                                "description": "",
-                                "is_source_dest_checked": true,
-                                "link_nic": [],
-                                "link_public_ip": [],
-                                "mac_address": "########",
-                                "net_id": "##id-4##",
-                                "nic_id": "##id-12##",
-                                "private_dns_name": "########",
-                                "private_ips": [
-                                    {
-                                        "is_primary": true,
-                                        "link_public_ip": [],
-                                        "private_dns_name": "########",
-                                        "private_ip": "########"
-                                    }
-                                ],
-                                "security_groups": [
-                                    {
-                                        "security_group_id": "##id-6##",
-                                        "security_group_name": "Terraform-SG-TF113"
-                                    }
-                                ],
-                                "state": "available",
-                                "subnet_id": "##id-13##",
-                                "subregion_name": "us-east-2b",
-                                "tags": [
+                                    },
                                     {
                                         "key": "Key:",
                                         "value": ":value-tags"
-                                    },
-                                    {
-                                        "key": "Key-2",
-                                        "value": "value-tags-2"
                                     }
                                 ]
                             },
@@ -588,44 +626,6 @@
                                         "value": "Nic-2"
                                     }
                                 ]
-                            },
-                            {
-                                "account_id": "##id-1##",
-                                "description": "Primary network interface",
-                                "is_source_dest_checked": true,
-                                "link_nic": [
-                                    {
-                                        "delete_on_vm_deletion": true,
-                                        "device_number": 0,
-                                        "link_nic_id": "##id-9##",
-                                        "state": "attached",
-                                        "vm_account_id": "##id-1##",
-                                        "vm_id": "##id-3##"
-                                    }
-                                ],
-                                "link_public_ip": [],
-                                "mac_address": "########",
-                                "net_id": "##id-4##",
-                                "nic_id": "##id-10##",
-                                "private_dns_name": "########",
-                                "private_ips": [
-                                    {
-                                        "is_primary": true,
-                                        "link_public_ip": [],
-                                        "private_dns_name": "########",
-                                        "private_ip": "########"
-                                    }
-                                ],
-                                "security_groups": [
-                                    {
-                                        "security_group_id": "##id-6##",
-                                        "security_group_name": "Terraform-SG-TF113"
-                                    }
-                                ],
-                                "state": "in-use",
-                                "subnet_id": "##id-7##",
-                                "subregion_name": "us-east-2a",
-                                "tags": []
                             }
                         ],
                         "request_id": "########"
@@ -736,13 +736,13 @@
                         "private_ip": "########",
                         "private_ips": [
                             {
-                                "is_primary": true,
+                                "is_primary": false,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
                             },
                             {
-                                "is_primary": false,
+                                "is_primary": true,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
@@ -775,7 +775,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.security_group01",
@@ -806,13 +806,13 @@
                         "private_ip": "########",
                         "private_ips": [
                             {
-                                "is_primary": true,
+                                "is_primary": false,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
                             },
                             {
-                                "is_primary": false,
+                                "is_primary": true,
                                 "link_public_ip": [],
                                 "private_dns_name": "########",
                                 "private_ip": "########"
@@ -841,7 +841,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.security_group01",
@@ -905,7 +905,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_security_group.security_group01",

--- a/tests/qa_provider_oapi/data/nets/TF-145_vm_resource_private_with_multiple_NICs_attributes_ok/step1.vm_resource_private_with_multiple_NICs_ok.ref
+++ b/tests/qa_provider_oapi/data/nets/TF-145_vm_resource_private_with_multiple_NICs_attributes_ok/step1.vm_resource_private_with_multiple_NICs_ok.ref
@@ -96,7 +96,7 @@
                         "timeouts": null
                     },
                     "sensitive_attributes": [],
-                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwfX0=",
+                    "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjo2MDAwMDAwMDAwMDAsImRlbGV0ZSI6NjAwMDAwMDAwMDAwLCJyZWFkIjozMDAwMDAwMDAwMDAsInVwZGF0ZSI6MzAwMDAwMDAwMDAwfX0=",
                     "dependencies": [
                         "outscale_net.outscale_net",
                         "outscale_subnet.outscale_subnet"

--- a/utils/utils.go
+++ b/utils/utils.go
@@ -37,7 +37,7 @@ const (
 	ReadDefaultTimeout   time.Duration = 5 * time.Minute
 	UpdateDefaultTimeout time.Duration = 10 * time.Minute
 	DeleteDefaultTimeout time.Duration = 5 * time.Minute
-	TestAccVmType        string        = "tinav7.c2r2p2"
+	TestAccVmType        string        = "tinav6.c2r2p2"
 	LinkedPolicyNotFound string        = "5102"
 	InvalidState         string        = "InvalidState"
 	SuffixConfigFilePath string        = "/.osc/config.json"
@@ -389,4 +389,11 @@ func CheckDateFormat(dateFormat string) error {
 		return fmt.Errorf(" Expiration date: '%s' should be after current date '%s'", settingDate, currentDate)
 	}
 	return nil
+}
+
+func GetTypeSetDifferencesForUpdating(oldTypeSet, newTypeSet *schema.Set) (*schema.Set, *schema.Set) {
+	inter := oldTypeSet.Intersection(newTypeSet)
+	toAdd := newTypeSet.Difference(inter)
+	toRemove := oldTypeSet.Difference(inter)
+	return toRemove, toAdd
 }

